mmv1/third_party/terraform/.teamcity/components/builds/build_steps.kt

```diff
@@ -131,4 +131,44 @@ fun BuildSteps.runAcceptanceTests() {
             """.trimIndent()
         })
     }
+}
+
+fun BuildSteps.saveArtifactsToGCS() {
+    step(ScriptBuildStep {
+        name = "Tasks after running nightly tests: push artifacts(debug logs) to GCS"
+        scriptContent = """
+            #!/bin/bash
+            echo "Post-test step - storge artifacts(debug logs) to GCS"
+
+            # Authenticate gcloud CLI
+            echo "${'$'}{GOOGLE_CREDENTIALS}" > google-account.json
+            chmod 600 google-account.json
+            gcloud auth activate-service-account --key-file=google-account.json
+
+            # Detect Trigger Method 
+            TRIGGERED_BY_USERNAME=%teamcity.build.triggeredBy.username%
+            BRANCH_NAME=%teamcity.build.branch%
+            if [[ "${'$'}TRIGGERED_BY_USERNAME" = "n/a" ]] ; then
+                echo "Build was triggered as part of automated testing. We know this because the `triggeredBy.username` value was `n/a`, value: ${'$'}{TRIGGERED_BY_USERNAME}"
+                FOLDER="nightly"
+            else
+                echo "Build was triggered manually. We know this because `triggeredBy.username` has a non- `n/a` value: ${'$'}{TRIGGERED_BY_USERNAME}"
+                FOLDER="manual/${'$'}BRANCH_NAME"
+            fi
+
+
+            # Copy logs to GCS
```

Comment:
Am I right assuming that the `teamcity-nightly-logs` bucket is intended to contain all the logs from cron- and manually-triggered tests that use the main branch of the downstream provider repos only? I.e. any tests run against the MM upstream forks using the VCR projects will be excluded?

If that's correct then we'd need to add some logic to make this build step be conditionally included only in the nightly projects.

Reply:
Some context: the code in `.teamcity/components/builds/build_configuration_per_package.kt` is used to create the build configurations in the `Upstream MM Testing` and `Nightly Tests` sub projects within Google and Google Beta projects in TeamCity. Inclusion of the new `saveArtifactsToGCS()` function would either need to be made conditional or the code restructured somehow

Reply:
> My original plan was to only store the logs from the cron-triggered nightly tests.

That could be achieved by inspecting [if a build is tagged to indicate it's cron triggered](https://github.com/hashicorp/magic-modules/blob/6681336af2f445057526fd050033546ee6b87502/mmv1/third_party/terraform/.teamcity/components/builds/build_steps.kt#L47-L51)? If that was done then you could have the build step present across `Upstream MM Testing` and `Nightly Tests` but it only performs the upload step when builds are cron triggered (and no non-sweepers are triggered by CRON in `Upstream MM Testing`). Conditionally skipping steps is [already present in the code base](https://github.com/hashicorp/magic-modules/blob/6681336af2f445057526fd050033546ee6b87502/mmv1/third_party/terraform/.teamcity/components/builds/build_steps.kt#L117-L127) in [a few places](https://github.com/hashicorp/magic-modules/blob/6681336af2f445057526fd050033546ee6b87502/mmv1/third_party/terraform/.teamcity/components/builds/vcr_build_steps.kt#L100-L103), so there is precedent, but I'd argue that any conditional behaviour has to be made explicit!

Reply:
> Now I see that I might not set up the GCS path correctly to differentiate between these various projects/purposes, I'll modify it to make sure that they'll be stored in dedicated locations 

Yeah, there's a bit of a trap of figuring what the project boundaries mean! How you've included the project id in the prefix in 0eeab5bc34cc9a2385f72cf250017ddf3000b0d3 looks good - the ID won't change and indicates the provider and purpose pretty well. Purpose gets a bit muddy when there are cron versus manual triggered builds in the nightly test projects though 😬 

>Perhaps renaming the bucket to something like teamcity-logs would be less confusing. What do you think?

I think that change would be good - especially if MM Upstream builds will end up storing logs there. The project id in the object prefix will help to distinguish nightly tests from MM Upstream stuff, if you decide to include them